core/rpc_client.py

```diff
@@ -0,0 +1,224 @@
+#!/usr/bin/env python
+#  -*- coding: utf-8 -*-
+# *********************************************************************
+# plankton - a library for creating hardware device simulators
+# Copyright (C) 2016 European Spallation Source ERIC
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# *********************************************************************
+
+import zmq
+import uuid
+import types
+
+try:
+    import exceptions
+except ImportError:
+    import builtins as exceptions
+
+
+class JSONRPCServerSideException(Exception):
+    def __init__(self, type, message):
+        """
+        This exception type replaces exceptions that are raised on the server,
+        but unknown (i.e. not in the exceptions-module) on the client side.
+        To retain as much information as possible, the exception type on the server and
+        the message are stored.
+
+        :param type: Type of the exception on the server side.
+        :param message: Exception message.
+        """
+        self.type = type
+        self.message = message
+
+    def __str__(self):
+        return 'Exception on server side of type \'{}\': \'{}\''.format(self.type, self.message)
+
+
+class JSONRPCProtocolException(Exception):
+    """
+    An exception type for exceptions related to the transport protocol, i.e.
+    malformed requests etc.
+    """
+    pass
+
+
+class ZMQJSONRPCConnection(object):
+    def __init__(self, host='127.0.0.1', port='10000'):
+        self._socket = self._get_zmq_req_socket()
+        self._socket.connect('tcp://{0}:{1}'.format(host, port))
+
+    def _get_zmq_req_socket(self):
+        context = zmq.Context()
+        return context.socket(zmq.REQ)
+
+    def json_rpc(self, method, *args):
+        """
+        This method takes a ZMQ REQ-socket and submits a JSON-object containing
+        the RPC (JSON-RPC 2.0 format) to the supplied method with the supplied arguments.
+        Then it waits for a reply from the server and blocks until it has received
+        a JSON-response. The method returns the response and the id it used to tag
+        the original request, which is a random UUID (uuid.uuid4).
+
+        :param socket: ZMQ REQ-socket.
```

Comment:
Yes...I wrote the docstrings too early I suppose and then didn't check properly. Thanks.